File: ios/Sources/BushaPay/Internal/DeepLink.swift
import Foundation

enum DeepLink {
    /// Returns the Busha app's URL scheme for `env` on iOS.
    static func bushaAppScheme(_ env: BushaEnvironment) -> String {
        switch env {
        case .sandbox:
            return "co.busha.boro.development"
        case .live:
            return "co.busha.apple"
        }
    }

    /// Builds the Busha app deep link.
    ///
    /// `reference` is included only when set, mirroring the Flutter and RN
    /// SDKs' shape.
    static func buildBushaAppDeepLink(
        scheme: String,
        config: BushaPayConfig,
        publicKey: String,
        callbackUrl: String
    ) -> URL? {
        var components = URLComponents()
        components.scheme = scheme
        components.host = "busha.co"
        components.path = "/pay"
        var items: [URLQueryItem] = [
            URLQueryItem(name: "public_key", value: publicKey),
            URLQueryItem(name: "quote_amount", value: config.quoteAmount),
            URLQueryItem(name: "quote_currency", value: config.quoteCurrency),
            URLQueryItem(name: "target_currency", value: config.targetCurrency),
        ]
        if let reference = config.reference {
            items.append(URLQueryItem(name: "reference", value: reference))
        }
        items.append(URLQueryItem(name: "callback_url", value: callbackUrl))
        components.queryItems = items
        return components.url
    }
}

/// URI schemes that should stay inside the WebView.
let kWebSchemes: Set<String> = ["http", "https", "about", "data", "blob"]

func isWebScheme(_ scheme: String) -> Bool {
    kWebSchemes.contains(scheme.lowercased())
}

File: android/pay-android/src/main/kotlin/co/busha/pay/internal/DeepLink.kt
package co.busha.pay.internal

import co.busha.pay.BushaEnvironment
import co.busha.pay.BushaPayCancelledReason
import co.busha.pay.BushaPayConfig
import co.busha.pay.BushaPayResult
import java.net.URLEncoder

/** URI schemes that should stay inside the WebView. */
internal val WEB_SCHEMES = setOf("http", "https", "about", "data", "blob")

internal fun isWebScheme(scheme: String): Boolean =
    WEB_SCHEMES.contains(scheme.lowercase())

/** The Busha app's URL scheme for [env] on Android. */
internal fun bushaAppScheme(env: BushaEnvironment): String = when (env) {
    BushaEnvironment.SANDBOX -> "co.busha.android.development"
    BushaEnvironment.LIVE -> "co.busha.android"
}

/**
 * Builds the Busha app deep link. `reference` is included only when set,
 * mirroring the Flutter, iOS, and React Native SDKs.
 */
internal fun buildBushaAppDeepLink(
    scheme: String,
    config: BushaPayConfig,
    publicKey: String,
    callbackUrl: String,
): String {
    val params = LinkedHashMap<String, String>()
    params["public_key"] = publicKey
    params["quote_amount"] = config.quoteAmount
    params["quote_currency"] = config.quoteCurrency
    params["target_currency"] = config.targetCurrency
    config.reference?.let { params["reference"] = it }
    params["callback_url"] = callbackUrl
    val query = params.entries.joinToString("&") { "${encode(it.key)}=${encode(it.value)}" }
    return "$scheme://busha.co/pay?$query"
}

/**
 * Parses a callback deep link into a result.
 *
 * `status=cancelled` maps to a [BushaPayCancelledReason.REJECTED]
 * cancellation — the Busha app reports an explicit rejection and hands
 * back the request id. A resume-without-callback is `ABANDONED` instead,
 * and is decided by the launcher, not here.
 */
internal fun parseCallback(url: String): BushaPayResult {
    val query = parseUrl(url)?.query ?: emptyMap()
    val status = query["status"]
    val paymentRequestId = query["paymentRequestId"] ?: ""
    return when (status) {
        "completed" -> successFromCallback(paymentRequestId)
        "cancelled" -> BushaPayResult.Cancelled(
            reason = BushaPayCancelledReason.REJECTED,
            paymentId = paymentRequestId,
        )
        else -> {
            val code = query["error_code"] ?: status ?: "unknown"
            val message = query["error_message"] ?: "Payment failed"
            BushaPayResult.Error(message = message, code = code)
        }
    }
}

private fun encode(s: String): String =
    URLEncoder.encode(s, "UTF-8").replace("+", "%20")
